### PR TITLE
Optimize PTX gated loads index calculation

### PR DESCRIPTION
**Pull request by SzymonOzog**
Transforms 
```
     add.s32         %alu_s32_46, %alu_s32_18, %const_s32_18;
     rem.s32         %alu_s32_47, %alu_s32_46, %const_s32_6;
     mul.lo.s32      %alu_s32_48, %alu_s32_47, %const_s32_23;
     cvt.u64.s32     %cast_u64_12, %alu_s32_48;
     add.u64         %alu_u64_12, %dat_u64_1, %cast_u64_12;
     @%alu_pred_21   ld.global.f32 %val_f32_28, [%alu_u64_12+0];
     @!%alu_pred_21  mov.b32 %val_f32_28, %const_f32_0;
```
into: 
```
    @%alu_pred_28   bra $if_27;
    mov.b32         %const_s32_56, 0;
    add.s32         %alu_s32_46, %alu_s32_18, %const_s32_17;
    rem.s32         %alu_s32_47, %alu_s32_46, %const_s32_5;
    mul.lo.s32      %alu_s32_48, %alu_s32_47, %const_s32_22;
    cvt.u64.s32     %cast_u64_12, %alu_s32_48;
    add.u64         %alu_u64_12, %dat_u64_1, %cast_u64_12;
    ld.global.f32   %val_f32_28, [%alu_u64_12+0];
    @!%alu_pred_28  bra $if_27_true;
$if_27:
    mov.b32         %val_f32_28, %const_f32_0;
$if_27_true:

```
and takes `speed_compare_ptx_cuda` to 0.97. But no gains on resnet - will look into it now

### CODE REVIEW

**Comment by geohot** [comment]
Afaik this doesn't work, the loaded address may be invalid and it may page fault. What is CUDA generating?

**Comment by SzymonOzog** [comment]
> Afaik this doesn't work, the loaded address may be invalid and it may page fault. What is CUDA generating?

oh  `@%alu_pred_28   bra $if_27;` checks the index, CUDA does the same thing but it always moves 0 into the register, PTX only does that if the load will not be valid. This is the cuda kernel:
```
     mov.f32         %f165, 0f00000000;
     mov.f32         %f161, %f165;
     mov.f32         %f162, %f165;
     mov.f32         %f163, %f165;
     mov.f32         %f164, %f165;
     @%p22 bra       $L__BB0_13;

     ld.global.f32   %f162, [%rd9];
     ld.global.f32   %f163, [%rd9+3584];
     ld.global.f32   %f161, [%rd9+7168];
     add.s32         %r90, %r108, -894;
     mul.hi.s32      %r91, %r90, -1840700269;
     add.s32         %r92, %r91, %r90;
     shr.u32         %r93, %r92, 31;
     shr.s32         %r94, %r92, 11;
     add.s32         %r95, %r94, %r93;
     mul.lo.s32      %r96, %r95, 3584;
     sub.s32         %r97, %r90, %r96;
     mul.wide.s32    %rd33, %r97, 4;
     add.s64         %rd34, %rd2, %rd33;
     ld.global.f32   %f164, [%rd34];

 $L__BB0_13:
```

**Comment by geohot** [comment]
Oh, I missed the initial jump over the load.

Is there a good way to test this?

**Comment by SzymonOzog** [comment]
> Is there a good way to test this?

I've added the test, it's not perfect since it needs to include the function but it works

File: tinygrad/renderer/assembly.py [hunk 1]
```diff
@@ -133,14 +145,21 @@ def _cast(a, dtype:DType, atype:DType, bitcast=False, u=None, pred=False):
...
       kk(lang.asm_for_op[BinaryOps.ADD](r[vin[0]], r[vin[0]], "1", dtypes.int, lang.types[dtypes.int]),
           lang.asm_for_op[BinaryOps.CMPLT](pred:=ssa("pred", dtype="pred"), r[vin[0]], r[vin[0].vin[1]], dtypes.int, lang.types[dtypes.int]))
-      kk(*lang.render_bra(r_label[vin[0]], pred, f"{r_label[vin[0]]}_exit"), f"{r_label[vin[0]]}_exit:")
-    elif uop is UOps.ENDIF:
+      kk(*lang.render_bra(r_label[vin[0]], pred))
+    elif uop == UOps.ENDIF:
+      kk(f"@!{_cast(r[vin[0].vin[0]], dtypes.bool, vin[0].vin[0].dtype, u=u, pred=True)} bra {r_label[vin[0]]}_true;")
       kk(f"{r_label[vin[0]]}:")
+      if args:
+        if args[0].dtype.count > 1:
```
Comment (geohot) [inline tinygrad/renderer/assembly.py:158]: Why does this need to be an if statement? You can always call scalar on the dtype.

File: tinygrad/renderer/assembly.py [hunk 2]
```diff
@@ -263,14 +278,13 @@ def render_local(self, dest, name, size, dtype) -> List[str]:
 
   def render_loop(self, idx, start, label, acc=None) -> List[str]: return [f"mov.u32 {idx}, {start};", f"{label}:"]
 
-  def render_bra(self, b1, pred=None, b2=None) -> List[str]: return [f"@{pred} bra {b1};", f"@!{pred} bra {b2};"] if pred else [f"bra {b1};"]
+  def render_bra(self, b1, pred=None) -> List[str]: return [f"@{pred} bra {b1};"] if pred else [f"bra {b1};"]
 
   def mem_type(self, dtype): return 's8' if dtype.itemsize == 1 else 'b16' if dtype == dtypes.float16 else self.types[dtype]
 
   def render_load(self, loc, dest, dtype, gate=None, alt=None, ss="", offset=0) -> List[str]:
```
Comment (geohot) [inline tinygrad/renderer/assembly.py:285]: What does gate do now, this function looks broken. Does it need to have its signature changed?

Comment (SzymonOzog) [reply tinygrad/renderer/assembly.py:285]: Oh previously it generated 
```
 95     @%pred_pred_0   bra $loop_0;
 96     @!%pred_pred_0  bra $loop_0_exit;
 97 $loop_0_exit:
```
The second jump does nothing so I just removed that - never needed that b2 in the first place

File: tinygrad/renderer/assembly.py (hunk 1 above)
Comment (SzymonOzog) [reply tinygrad/renderer/assembly.py:158]: r[uop] contains a list if the count is bigger than 1 so we need to mov to all associated registers then

**Comment by geohot** [comment]
There's still unused parameters in functions

**Comment by SzymonOzog** [comment]
> There's still unused parameters in functions

Ohh sorry, misread your previous comment - updated the signatures now

**Pull request merged**